[OpenXR] Use a WorkQueue for the render loop

carlosgcampos opened the pull request:
#### 356f67eae83172534ddf23df2a69a017fbd05978
<pre>
[OpenXR] Use a WorkQueue for the render loop
<a href="https://bugs.webkit.org/show_bug.cgi?id=297451">https://bugs.webkit.org/show_bug.cgi?id=297451</a>

Reviewed by Fujii Hironori.

We currently use a thread to poll the events and process the frame, but
several objects are created in main thread and used from both threads or
can be released in both threads too. Using a WorkQueue we can move all
the session handling to the render thread and simplify the render loop.

Canonical link: <a href="https://commits.webkit.org/298948@main">https://commits.webkit.org/298948@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8db30a34241681630fdeff38fc80d48c2fba327d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117277 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36947 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27560 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123376 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69262 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37643 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45534 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88995 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120210 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29956 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105148 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69503 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29017 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67049 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99355 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23444 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126497 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44174 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33185 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97665 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44530 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101384 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97460 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42811 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20740 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18713 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44047 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49706 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43503 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46848 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45199 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/20ecbdcd40aacd2f070bc407548184e5d5661e4b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49448

spenap commented:
Looking very nice. I'll check it locally after I rebase the API test against it. I saw you already move the didStart callback to the right place

spenap commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -253,9 +258,9 @@ void OpenXRCoordinator::endSessionIfExists(std::optional<WebCore::PageIdentifier
                 sessionEventClient->sessionDidEnd(m_deviceIdentifier);
             }
 
+            page.uiClient().didEndXRSession(page);
```
I see you moved this before we clear `m_state`. Was this so that if the client were to query something back,`m_state` would still have the previous information? Do we allow that kind of re-entrant behaviour?

spenap commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:
```diff
@@ -60,60 +60,57 @@ class OpenXRCoordinator final : public PlatformXRCoordinator {
-    PollResult pollEvents(Box<RenderState>);
-    PlatformXR::FrameData populateFrameData(Box<RenderState>);
+    PollResult pollEvents();
+    std::unique_ptr<OpenXRSwapchain> createSwapchain(uint32_t width, uint32_t height, bool alpha);
```
This can be made `const` 
```suggestion
    std::unique_ptr<OpenXRSwapchain> createSwapchain(uint32_t width, uint32_t height, bool alpha) const;
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -171,19 +167,25 @@ std::unique_ptr<OpenXRSwapchain> OpenXRCoordinator::createSwapchain(uint32_t wid
+    WTF::switchOn(m_state,
+        [&](Idle&) { },
+        [&](Active& active) {
+            active.renderQueue->dispatch([this, width, height, alpha] {
```
No, the coordinator is not refcounted because it's a never destroyed singleton

fujii commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -38,14 +38,13 @@
 
 struct OpenXRCoordinator::RenderState {
     std::atomic<bool> terminateRequested;
+    bool pendingFrame;
```
I think it's better to have an initializer.
> bool pendingFrame { false };

carlosgcampos committed:
[OpenXR] Use a WorkQueue for the render loop
https://bugs.webkit.org/show_bug.cgi?id=297451

Reviewed by Fujii Hironori.

We currently use a thread to poll the events and process the frame, but
several objects are created in main thread and used from both threads or
can be released in both threads too. Using a WorkQueue we can move all
the session handling to the render thread and simplify the render loop.

Canonical link: https://commits.webkit.org/298948@main
